### PR TITLE
Fixes part of #59: Introduce app module Robolectric tests + Bazel [BLOCKED: #1622]

### PR DESCRIPTION
**Pull request by miaboloix**
## Explanation
Fixes part of #59: Introduce app module Robolectric tests + Bazel [BLOCKED: #1622]

This PR sets up Robolectric tests for app module + Bazel. Because `rules_kotlin` does not have a `kt_android_local_test` rule, each test must be built in it's own `kt_android_library` and fed into an `android_local_test` rule as a dependency - the macro file `app_tests.bzl` handles this. On top of this, certain necessary data binding processing is not being handled by `kt_android_library` and `android_local_test`, causing `DataBinderMapperImpl.java` not to be generated. The workaround for this is manually creating `DataBinderMapperImpl.java` and adding it to the source file list for each test target. 

In order to set up the correct version of Bazel, please follow the instructions laid out in [this document](https://docs.google.com/document/d/1V0RMehfi3GLVsZ6dyin1nTo4_TO28LACp4wPlEnvnM0/edit?usp=sharing). 

To build and run all the tests in app module, run 
```
~/oppia-bazel/bazel-bin/src/bazel test //app/... --android_databinding_use_v3_4_args --experimental_android_databinding_v2 --override_repository=android_tools=/tmp/android_tools --subcommands --java_header_compilation=false --noincremental_dexing --define=android_standalone_dexing_tool=d8_compat_dx
```
To build and run a specific test in app module, run 
```
~/oppia-bazel/bazel-bin/src/bazel test //app:<name_of_test> --android_databinding_use_v3_4_args --experimental_android_databinding_v2 --override_repository=android_tools=/tmp/android_tools --subcommands --java_header_compilation=false --noincremental_dexing --define=android_standalone_dexing_tool=d8_compat_dx
```

## Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [ ] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [ ] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [ ] The PR is made from a branch that's **not** called "develop".
- [ ] The PR is made from a branch that is up-to-date with "develop".
- [ ] The PR's branch is based on "develop" and not on any other branch.
- [ ] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

**Comment by BenHenning** [comment]
@miaboloix this is in a draft state--is it ready for full review?

**Comment by miaboloix** [comment]
> @miaboloix this is in a draft state--is it ready for full review?

Yes - it is ready!

File: app/BUILD.bazel [hunk 1]
```diff
@@ -513,10 +514,773 @@ kt_android_library(
         artifact("androidx.viewpager:viewpager:1.0.0"),
         artifact("com.caverock:androidsvg-aar"),
         artifact("javax.annotation:javax.annotation-api:jar"),
+        artifact("androidx.navigation:navigation-fragment:2.0.0"),
```
Comment (BenHenning) [inline app/BUILD.bazel]: Why weren’t these needed before adding the tests?

File: app/BUILD.bazel [hunk 2]
```diff
@@ -513,10 +514,773 @@ kt_android_library(
...
+        artifact("androidx.navigation:navigation-fragment-ktx:2.0.0"),
+        artifact("androidx.navigation:navigation-ui:2.0.0"),
+        artifact("androidx.navigation:navigation-ui-ktx:2.0.0"),
     ],
     visibility = ["//visibility:public"],
 )
 
+kt_android_library(
```
Comment (BenHenning) [inline app/BUILD.bazel:600]: Awesome idea. Are these being properly kept out of the binary build?

Also, test only libraries should be marked as test_only

Comment (BenHenning) [inline app/BUILD.bazel:600]: Please add a brief line comment providing an explanation what this should contain to provide guidance to team members one when the library should be added to.

File: app/BUILD.bazel [hunk 3]
```diff
@@ -513,10 +514,773 @@ kt_android_library(
...
+        "//testing",
+        artifact("org.mockito:mockito-core"),
+    ]
+)
+
+TEST_DEPS = [
+    ":app",
+    ":dagger",
+    "//utility",
+    "//testing",
```
Comment (BenHenning) [inline app/BUILD.bazel]: Please sort.

File: app/BUILD.bazel [hunk 4]
```diff
@@ -513,10 +514,773 @@ kt_android_library(
...
+    artifact("androidx.test:runner:1.2.0"),
+    artifact("org.jetbrains.kotlinx:kotlinx-coroutines-test:1.2.2"),
+    artifact("org.robolectric:annotations:4.3"),
+    artifact("org.mockito:mockito-core"),
+    artifact("androidx.test.espresso:espresso-contrib:3.1.0"),
+    artifact("androidx.test.espresso:espresso-core:3.2.0"),
+    artifact("androidx.test.espresso:espresso-intents:3.1.0"),
+]
+
+ALTERED_TESTS = [
```
Comment (BenHenning) [inline app/BUILD.bazel]: Needs documentation. What are these files? Why are they here? How do I know when to add another?

Also I missed this earlier, but private members of a Bazel file should start with an underscore. Ditto for all other file lists and related. 

You may want to spruce up the documentation for other libraries and lists to provide similar pointers for team members.

File: app/BUILD.bazel [hunk 5]
```diff
@@ -513,10 +514,773 @@ kt_android_library(
...
+    name = "HomeActivityLocalTest",
+    srcs = [
+        "src/test/java/org/oppia/app/home/HomeActivityLocalTest.kt",
+        "src/test/java/DataBinderMapperImpl.java"
+    ],
+    test_class = "org.oppia.app.home.HomeActivityLocalTest",
+    deps = TEST_DEPS,
+)
+
+# TODO: Test doesn't pass
```
Comment (BenHenning) [inline app/BUILD.bazel]: Here and elsewhere: I think you can use #973 as the reference here, but please confirm the failing tests match the unresolved checkboxes in that issue.

File: app/BUILD.bazel [hunk 6]
```diff
@@ -513,10 +514,773 @@ kt_android_library(
...
+app_test(
+    name = "ExplorationActivityLocalTest",
+    srcs = [
+        "src/test/java/org/oppia/app/player/exploration/ExplorationActivityLocalTest.kt",
+        "src/test/java/DataBinderMapperImpl.java"
```
Comment (BenHenning) [inline app/BUILD.bazel]: If this is needed for all app module tests, let’s temporarily add it directly to the app_test macro. We should also have a TODO to remove this once we fix databinding in android_local_test.

File: app/BUILD.bazel [hunk 7]
```diff
@@ -513,10 +514,773 @@ kt_android_library(
...
+    ],
+    test_class = "org.oppia.app.settings.profile.ProfileRenameActivityTest",
+    deps = TEST_DEPS,
+)
+
+app_test(
+    name = "ProfileResetPinActivityTest",
+    srcs = [
+        "update_src/sharedTest/java/org/oppia/app/settings/profile/ProfileResetPinActivityTest",
```
Comment (BenHenning) [inline app/BUILD.bazel]: It seems a bit weird that these updated files don’t end in kt. Any particular reason why they don’t?

File: app/app_test.bzl [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  '''
+
+  oppia_android_test(
+    name = name,
+    srcs = srcs,
+    custom_package = "org.oppia.app.test",
+    test_class = test_class,
+    test_manifest = "src/test/AndroidManifest.xml",
+    deps = deps,
+    enable_data_binding = 1,
```
Comment (BenHenning) [inline app/app_test.bzl]: Prefer to use True instead of 1.

File: app/build.gradle [hunk 9]
```diff
@@ -61,6 +61,7 @@ android {
     test {
       java.srcDirs += "src/sharedTest/java"
       kotlin.srcDirs += "src/sharedTest/java"
+      java.exclude('**/DataBinderMapperImpl.java')
```
Comment (BenHenning) [inline app/build.gradle]: Please add a comment to explain and provide context on why this is being done. Otherwise people might accidentally change it without understanding the nuance that you do.

File: app/src/sharedTest/java/org/oppia/app/administratorcontrols/AppVersionActivityTest.kt [hunk 10]
```diff
@@ -79,7 +78,7 @@ class AppVersionActivityTest {
         withText(
           String.format(
             context.resources.getString(R.string.app_version_name),
-            BuildConfig.VERSION_NAME
+            context.packageManager.getPackageInfo(context.packageName, 0).versionName
```
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/administratorcontrols/AppVersionActivityTest.kt]: Maybe should move this to an extension function on package manager so that it can be reused everywhere that we need it. What do you think?

File: app/src/sharedTest/java/org/oppia/app/utility/DrawableMatcher.kt [hunk 11]
```diff
@@ -10,11 +10,12 @@ import org.hamcrest.TypeSafeMatcher
...
 // https://github.com/dbottillo/Blog/blob/espresso_match_imageview/app/src/androidTest/java/com/danielebottillo/blog/config/DrawableMatcher.java
 /** This class mainly provides a custom matcher to test whether the drawable-image is correctly shown in ImageView. */
-class DrawableMatcher internal constructor(
+// TODO(#1655): Re-restrict access to fields in tests post-Gradle.
+class DrawableMatcher constructor(
   private val expectedId: Int
 ) : TypeSafeMatcher<View>(View::class.java) {
   companion object {
-    internal const val NONE = -1
+    const val NONE = -1
```
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/utility/DrawableMatcher.kt:17]: This actually seems like something that shouldn’t have been internal to begin with since it’s part of this class’s API. Ditto for ANY.

File: app/src/test/java/DataBinderMapperImpl.java [hunk 12]
```diff
@@ -0,0 +1,7 @@
+package androidx.databinding;
+
+public class DataBinderMapperImpl extends MergedDataBinderMapper {
```
Comment (BenHenning) [inline app/src/test/java/DataBinderMapperImpl.java]: public final class

Also, definitely add documentation explaining why this is needed and a TODO to remove it once android_local_test properly supports data binding.

File: domain/BUILD.bazel [hunk 13]
```diff
@@ -150,13 +150,6 @@ domain_test(
     deps = TEST_DEPS,
 )
 
-domain_test(
-    name = "OnboardingFlowControllerTest",
```
Comment (BenHenning) [inline domain/BUILD.bazel:154]: Why was this removed? I believe the test was renamed.

File: oppia_android_test.bzl [hunk 14]
```diff
@@ -1,8 +1,8 @@
 load("@rules_jvm_external//:defs.bzl", "artifact")
 load("@io_bazel_rules_kotlin//kotlin:kotlin.bzl", "kt_android_library")
 
-def oppia_android_test(name, srcs, test_manifest, custom_package, test_class, deps, assets=None,
-                       assets_dir=None):
+def oppia_android_test(name, srcs, test_manifest, custom_package, test_class, deps,
+                       enable_data_binding=0, assets=None, assets_dir=None):
```
Comment (BenHenning) [inline oppia_android_test.bzl]: False rather than 0

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @miaboloix! Left some thoughts.

File: domain/BUILD.bazel (hunk 13 above)
Comment (miaboloix) [reply domain/BUILD.bazel:154]: Oh - I will change this, then!

File: oppia_android_test.bzl (hunk 14 above)
Comment (miaboloix) [reply oppia_android_test.bzl]: Done

File: app/build.gradle (hunk 9 above)
Comment (miaboloix) [reply app/build.gradle]: Done! Let me know if it needs a more in-depth explanation here - I figured the documentation in the actual file might provide this.

File: app/app_test.bzl (hunk 8 above)
Comment (miaboloix) [reply app/app_test.bzl]: Done

File: app/BUILD.bazel (hunk 2 above)
Comment (miaboloix) [reply app/BUILD.bazel:600]: I believe they are being properly kept out - the app library only builds source files. I will add test_only now.

File: app/BUILD.bazel (hunk 1 above)
Comment (miaboloix) [reply app/BUILD.bazel]: I believe because of Espresso.

File: app/BUILD.bazel (hunk 2 above)
Comment (miaboloix) [reply app/BUILD.bazel:600]: Done

File: app/BUILD.bazel (hunk 3 above)
Comment (miaboloix) [reply app/BUILD.bazel]: okay

File: app/BUILD.bazel (hunk 5 above)
Comment (miaboloix) [reply app/BUILD.bazel]: Sounds good and yes, I can confirm!

File: app/BUILD.bazel (hunk 7 above)
Comment (miaboloix) [reply app/BUILD.bazel]: I just figured that a target name shouldn't have the `.kt` notation - but in this particular case since it is being used as a source I understand how it could be strage. Should I go ahead and add the `.kt`?

File: app/src/test/java/DataBinderMapperImpl.java (hunk 12 above)
Comment (miaboloix) [reply app/src/test/java/DataBinderMapperImpl.java]: Done

File: app/src/sharedTest/java/org/oppia/app/utility/DrawableMatcher.kt (hunk 11 above)
Comment (miaboloix) [reply app/src/sharedTest/java/org/oppia/app/utility/DrawableMatcher.kt:17]: Okay

File: app/BUILD.bazel (hunk 6 above)
Comment (miaboloix) [reply app/BUILD.bazel]: Okay

File: app/src/sharedTest/java/org/oppia/app/administratorcontrols/AppVersionActivityTest.kt (hunk 10 above)
Comment (miaboloix) [reply app/src/sharedTest/java/org/oppia/app/administratorcontrols/AppVersionActivityTest.kt]: Do you mean create a function that would replace `.getPackageInfo(context.packageName, 0).versionName` ?

File: app/BUILD.bazel (hunk 4 above)
Comment (miaboloix) [reply app/BUILD.bazel]: Okay

File: WORKSPACE [hunk 15]
```diff
@@ -165,8 +166,13 @@ maven_install(
         "org.jetbrains.kotlinx:kotlinx-coroutines-test:1.2.2",
         "org.jetbrains:annotations:jar:13.0",
         "org.mockito:mockito-core:2.19.0",
+        "org.mockito:mockito-core:2.7.22",
         "org.robolectric:annotations:4.3",
         "org.robolectric:robolectric:4.3",
+        "androidx.navigation:navigation-fragment:2.0.0", #
```
Comment (anandwana001) [inline WORKSPACE]: Why we need `#` at the end?

**Review by anandwana001 — CHANGES_REQUESTED** [review]
Giving the first pass over this pr. 

@miaboloix  Also, could you describe the command to run all the test cases give in the pr description? 
I am getting in running this as well as applying the patch file give in one of your bazel databinding docs.

File: app/BUILD.bazel (hunk 1 above)
Comment (miaboloix) [reply app/BUILD.bazel]: Actually, now that I am looking at it - I think this isn't necessary. I will run all the tests and remove it if it isn't needed.

File: app/BUILD.bazel [hunk 16]
```diff
@@ -488,6 +500,7 @@ android_library(
     visibility = ["//visibility:private"],
 )
 
+# Final app library
```
Comment (BenHenning) [inline app/BUILD.bazel]: What does 'final' mean?

File: app/BUILD.bazel [hunk 17]
```diff
@@ -67,7 +67,10 @@ LISTENERS = [
 DATABINDING_LAYOUTS = ["src/main/res/layout*/**"]
 
 # TODO(#1617): Remove genrules post-gradle
-# Files altered by genrules to be excluded in the app library
+'''
+View model files to be processed by genrules. View models should be processed if they import
```
Comment (BenHenning) [inline app/BUILD.bazel]: Maybe frame this & others as: "View models that import resources or binding files must be included in this list"

File: app/BUILD.bazel [hunk 18]
```diff
@@ -510,125 +523,137 @@ kt_android_library(
...
-        artifact("androidx.navigation:navigation-ui:2.0.0"),
-        artifact("androidx.navigation:navigation-ui-ktx:2.0.0"),
     ],
     visibility = ["//visibility:public"],
 )
 
+'''
+Library that builds test-only source files. These files should not be included in the app library or
+the final binary but should be depedencies for tests.
+'''
```
Comment (BenHenning) [inline app/BUILD.bazel]:
```suggestion
Library that builds test-only source files. All app-specific files that are only used for testing purposes but are
not tests should be included in this list.  These files should not be included in the app library
the final binary but should be dependencies for tests.
'''
```

I suggest making this more instructional so people know how to update it.

File: app/BUILD.bazel [hunk 19]
```diff
@@ -510,125 +523,137 @@ kt_android_library(
...
+    artifact("androidx.test:runner:1.2.0"),
+    artifact("com.google.truth:truth"),
+    artifact("org.jetbrains.kotlin:kotlin-reflect"),
+    artifact("org.jetbrains.kotlinx:kotlinx-coroutines-test:1.2.2"),
+    artifact("org.mockito:mockito-core"),
+    artifact("org.robolectric:annotations:4.3"),
 ]
 
+'''
+Tests to be processed using the test genrule. Test files should be processed if they import resource
```
Comment (BenHenning) [inline app/BUILD.bazel]: Should this say that all test files should go here, instead?

File: app/BUILD.bazel [hunk 20]
```diff
@@ -642,145 +667,103 @@ account for altered package and file name.
  )
 for file in ALTERED_TESTS]
 
+# TODO(#973): Fix app module tests for Roboelectric.
```
Comment (BenHenning) [inline app/BUILD.bazel]: Can you actually add this at each test declaration for currently expected failing tests?

File: app/BUILD.bazel [hunk 21]
```diff
@@ -642,145 +667,103 @@ account for altered package and file name.
...
     deps = TEST_DEPS,
 )
 
 app_test(
     name = "StateFragmentAccessibilityTest",
-    srcs = [
-        "update_src/test/java/org/oppia/app/testing/player/state/StateFragmentAccessibilityTest",
-        "src/test/java/DataBinderMapperImpl.java",
-    ],
+    srcs = ["update_src/test/java/org/oppia/app/testing/player/state/StateFragmentAccessibilityTest"],
```
Comment (BenHenning) [inline app/BUILD.bazel]: I was hesitant to suggest this before, but now that I'm seeing we're mixing updated_src and not, I think this should utilize a macro to keep tests clearer, e.g.:

```bazel
app_test(
    name = "StateFragmentAccessibilityTest",
    srcs = [test_with_resources("src/test/java/org/oppia/app/testing/player/state/StateFragmentAccessibilityTest.kt")],
    ...
```

Where ``test_with_resources`` (or a better named macro) actually computes the updated_ + omit the kt extension in order to create the correct source reference. In this way, all tests properly refer to their exact source file, and we explicitly call out cases when tests are special & need to be in the updated test list above. This also avoids exposing the implementation specifics of the genrule.

File: app/src/test/java/DataBinderMapperImpl.java [hunk 22]
```diff
@@ -1,6 +1,13 @@
...
 
-public class DataBinderMapperImpl extends MergedDataBinderMapper {
+/**
+ * Generated code necessary to run tests with Bazel.
+ *
+ * Typically, this code is generated by Bazel and included in the srcs list when running tests.
+ * However, the `enable_data_binding` flag somehow affects this process. For now, this file must be
+ * included in the srcs list of any Bazel test in app module.
+ */
+public final class DataBinderMapperImpl extends MergedDataBinderMapper {
```
Comment (BenHenning) [inline app/src/test/java/DataBinderMapperImpl.java:10]: Please add TODO to remove on issue tracking making android_local_test work with data-binding.

File: app/src/sharedTest/java/org/oppia/app/utility/DrawableMatcher.kt (hunk 11 above)
Comment (BenHenning) [reply app/src/sharedTest/java/org/oppia/app/utility/DrawableMatcher.kt:17]: Reminder to remove the new TODO since access doesn't actually need to be re-restricted per the comment above. :)

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @miaboloix! Took another pass.

File: app/src/sharedTest/java/org/oppia/app/administratorcontrols/AppVersionActivityTest.kt (hunk 10 above)
Comment (BenHenning) [reply app/src/sharedTest/java/org/oppia/app/administratorcontrols/AppVersionActivityTest.kt]: Yes, something like:

```kotlin
fun PackageManager.getVersionName(): String {
  return getPackageInfo(...).versionName
}
```

Then we can do ``context.packageManager.getVersionName()`` where needed.

File: app/BUILD.bazel (hunk 7 above)
Comment (BenHenning) [reply app/BUILD.bazel]: Please see my newer comment & discuss there. I think if we can hide the implementation details of the genrule it should be fine.

File: app/build.gradle (hunk 9 above)
Comment (BenHenning) [reply app/build.gradle]: This looks good--thanks!

File: WORKSPACE (hunk 15 above)
Comment (miaboloix) [reply WORKSPACE]: We dont, removed them and alphabetized them :)

File: app/BUILD.bazel (hunk 16 above)
Comment (miaboloix) [reply app/BUILD.bazel]: I just meant that it was the 'final' library before building the binary - like the main library for the whole app. I can remove 'final', it is confusing.

File: app/BUILD.bazel (hunk 19 above)
Comment (miaboloix) [reply app/BUILD.bazel]: Not all test files are using the genrule, though. I am updating this documentation anyway (to be consistent with documentation I wrote a few PRs back) so please let me know if it is clearer now.

File: app/BUILD.bazel (hunk 20 above)
Comment (miaboloix) [reply app/BUILD.bazel]: Done

File: app/BUILD.bazel (hunk 18 above)
Comment (miaboloix) [reply app/BUILD.bazel]: Ok

File: app/BUILD.bazel (hunk 17 above)
Comment (miaboloix) [reply app/BUILD.bazel]: Leaving the new documentation from an older PR here - let me know if you would like me to change it.

File: app/src/test/java/DataBinderMapperImpl.java (hunk 22 above)
Comment (miaboloix) [reply app/src/test/java/DataBinderMapperImpl.java:10]: done

**Comment by miaboloix** [comment]
> Giving the first pass over this pr.
> 
> @miaboloix Also, could you describe the command to run all the test cases give in the pr description?
> I am getting in running this as well as applying the patch file give in one of your bazel databinding docs.

I linked to the document (it has been updated so that applying the patch is no longer needed) in the description as well as provided the correct syntax to run the tests. Let me know if you need anything else!

File: app/src/sharedTest/java/org/oppia/app/administratorcontrols/AppVersionActivityTest.kt (hunk 10 above)
Comment (miaboloix) [reply app/src/sharedTest/java/org/oppia/app/administratorcontrols/AppVersionActivityTest.kt]: Sure!

**Comment by anandwana001** [comment]
Putting this here for reference, after running all the test cases as per the command mentioned in the pr description. 
Screenshot 2020-08-21 at 04 02 12

**Review by anandwana001 — APPROVED** [review]
LGTM, Thanks @miaboloix

File: app/src/sharedTest/java/org/oppia/app/administratorcontrols/AppVersionActivityTest.kt [hunk 23]
```diff
@@ -149,6 +150,10 @@ class AppVersionActivityTest {
     }
   }
 
+  private fun PackageManager.getVersionName(): String {
+    return getPackageInfo(context.packageName, 0).versionName
```
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/administratorcontrols/AppVersionActivityTest.kt]:
```suggestion
    return getPackageInfo(context.packageName, /* flags= */ 0).versionName
```

To provide context to readers.

File: app/src/sharedTest/java/org/oppia/app/administratorcontrols/AppVersionActivityTest.kt [hunk 24]
```diff
@@ -78,7 +79,7 @@ class AppVersionActivityTest {
         withText(
           String.format(
             context.resources.getString(R.string.app_version_name),
-            context.packageManager.getPackageInfo(context.packageName, 0).versionName
+            context.packageManager.getVersionName()
```
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/administratorcontrols/AppVersionActivityTest.kt]: Is this used elsewhere in the codebase? If so, suggest moving the extension function to a Kotlin helper file in the utility package and importing it everywhere it's needed.

File: app/src/test/java/DataBinderMapperImpl.java [hunk 25]
```diff
@@ -1,5 +1,5 @@
 package androidx.databinding;
-
+//TODO(#1683): Remove when android_local_test works with data-binding.
```
Comment (BenHenning) [inline app/src/test/java/DataBinderMapperImpl.java]:
```suggestion
// TODO(#1683): Remove when android_local_test works with data-binding.
```

File: app/test_with_resources.bzl [hunk 26]
```diff
@@ -0,0 +1,20 @@
+def test_with_resources(file):
```
Comment (BenHenning) [inline app/test_with_resources.bzl:1]: Nice idea to combine the genrule & the target! This is a great simplification.

File: app/test_with_resources.bzl [hunk 27]
```diff
@@ -0,0 +1,20 @@
...
+  '''
+  Genrule for test files.
+
+  Because each databinding library must have a unique package name and manifest, resources must be
+  imported using the proper package name when building with Bazel. This genrule alters those imports
+  in order to keep Gradle building.
+  '''
+  native.genrule(
+    name = "update_" + file[0:-3],
+    srcs = [file[0:-3] + ".kt"],
```
Comment (BenHenning) [inline app/test_with_resources.bzl]: Could this just be ``srcs = [file]`` given we know it already has to be a Kotlin file?

File: app/BUILD.bazel (hunk 16 above)
Comment (BenHenning) [reply app/BUILD.bazel]: I think we should just be descriptive about what this represents, e.g.:

``Library containing all activity, fragment, and view-based UI flows in the app.`` or similar.

File: app/BUILD.bazel (hunk 19 above)
Comment (BenHenning) [reply app/BUILD.bazel]: Much clearer--thanks!

File: app/BUILD.bazel (hunk 21 above)
Comment (BenHenning) [reply app/BUILD.bazel]: This looks resolved now, and looks quite good with your additional simplification.

**Review by BenHenning — APPROVED** [review]
A few nits, but otherwise LGTM @miaboloix!

File: app/src/test/java/DataBinderMapperImpl.java (hunk 25 above)
Comment (miaboloix) [reply app/src/test/java/DataBinderMapperImpl.java]: Done

File: app/test_with_resources.bzl (hunk 26 above)
Comment (miaboloix) [reply app/test_with_resources.bzl:1]: Thank you!

File: app/test_with_resources.bzl (hunk 27 above)
Comment (miaboloix) [reply app/test_with_resources.bzl]: Fixed!

File: app/BUILD.bazel (hunk 16 above)
Comment (miaboloix) [reply app/BUILD.bazel]: Done

File: app/src/sharedTest/java/org/oppia/app/administratorcontrols/AppVersionActivityTest.kt (hunk 23 above)
Comment (miaboloix) [reply app/src/sharedTest/java/org/oppia/app/administratorcontrols/AppVersionActivityTest.kt]: Done

File: app/src/sharedTest/java/org/oppia/app/administratorcontrols/AppVersionActivityTest.kt (hunk 24 above)
Comment (miaboloix) [reply app/src/sharedTest/java/org/oppia/app/administratorcontrols/AppVersionActivityTest.kt]: Done!

Comment (miaboloix) [reply app/src/sharedTest/java/org/oppia/app/administratorcontrols/AppVersionActivityTest.kt]: I also implemented an extension function for `lastUpdateTime` because I saw it was needed just as much as `versionName`.

File: app/BUILD.bazel [hunk 28]
```diff
@@ -556,7 +556,7 @@ android_library(
     visibility = ["//visibility:private"],
 )
 
-# App library
+# Library containing all activity, fragment, and view-based UI flows in the app
```
Comment (BenHenning) [inline app/BUILD.bazel]: Nit: end with period (feel free to resolve once fixed).

File: app/src/main/java/org/oppia/app/utility/PackageManagerHelper.kt [hunk 29]
```diff
@@ -0,0 +1,20 @@
...
+
+import android.content.Context
+import android.content.pm.PackageManager
+import javax.inject.Inject
+
+/** Extension functions for PackageManager. */
+@Inject
+lateinit var context: Context
+
+fun PackageManager.getVersionName(): String {
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/utility/PackageManagerHelper.kt]: Nit: add KDoc here & below. Feel free to resolve once addressed.

File: app/src/main/java/org/oppia/app/utility/PackageManagerHelper.kt [hunk 30]
```diff
@@ -0,0 +1,20 @@
...
+  return getPackageInfo(context.packageName, /* flags=*/0).versionName
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/utility/PackageManagerHelper.kt]:
```suggestion
  return getPackageInfo(context.packageName, /* flags=*/ 0).versionName
```

Feel free to resolve once accepted.

File: app/src/main/java/org/oppia/app/utility/PackageManagerHelper.kt [hunk 31]
```diff
@@ -0,0 +1,20 @@
+package org.oppia.app.utility
+
+import android.content.Context
+import android.content.pm.PackageManager
+import javax.inject.Inject
+
+/** Extension functions for PackageManager. */
+@Inject
+lateinit var context: Context
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/utility/PackageManagerHelper.kt]: We aren't injecting anything, so this will always be null. Plus, we don't actually need this if we have access to PackageManager via extension.

File: app/src/main/java/org/oppia/app/utility/PackageManagerHelper.kt [hunk 32]
```diff
@@ -0,0 +1,20 @@
+package org.oppia.app.utility
+
+import android.content.Context
+import android.content.pm.PackageManager
+import javax.inject.Inject
+
+/** Extension functions for PackageManager. */
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/utility/PackageManagerHelper.kt]: Not actually documenting any elements, so suggest changing to non-doc comment.

File: app/src/sharedTest/java/org/oppia/app/administratorcontrols/AppVersionActivityTest.kt [hunk 33]
```diff
@@ -58,10 +59,7 @@ class AppVersionActivityTest {
     setUpTestApplicationComponent()
 
     val lastUpdateDateTime =
-      context.packageManager.getPackageInfo(
-        context.packageName,
-        /* flags= */ 0
-      ).lastUpdateTime
+      context.packageManager.getLastUpdateTime()
```
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/administratorcontrols/AppVersionActivityTest.kt]: Maybe fit this on one line?

**Review by BenHenning — CHANGES_REQUESTED** [review]
Removing 'approved' per the incorrect ``@Inject``.

File: app/src/sharedTest/java/org/oppia/app/administratorcontrols/AppVersionActivityTest.kt (hunk 33 above)
Comment (miaboloix) [reply app/src/sharedTest/java/org/oppia/app/administratorcontrols/AppVersionActivityTest.kt]: Done

File: app/src/main/java/org/oppia/app/utility/PackageManagerHelper.kt (hunk 32 above)
Comment (miaboloix) [reply app/src/main/java/org/oppia/app/utility/PackageManagerHelper.kt]: ok

File: app/src/main/java/org/oppia/app/utility/PackageManagerHelper.kt (hunk 31 above)
Comment (miaboloix) [reply app/src/main/java/org/oppia/app/utility/PackageManagerHelper.kt]: Ended up extending Context instead per our chat

File: app/src/main/java/org/oppia/app/utility/PackageManagerHelper.kt (hunk 29 above)
Comment (miaboloix) [reply app/src/main/java/org/oppia/app/utility/PackageManagerHelper.kt]: Okay

File: app/src/main/java/org/oppia/app/utility/ContextExtensions.kt [hunk 34]
```diff
@@ -0,0 +1,15 @@
+package org.oppia.app.utility
+
+import android.content.Context
+
+// Extension functions for Context that act as getters for PackageManager.
+
+/** Returns the current version name. */
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/utility/ContextExtensions.kt]:
```suggestion
/** Returns the current app version name. */
```

**Review by BenHenning — APPROVED** [review]
Thanks @miaboloix. LGTM!
